pkg/provisioner/ironic/ironic.go

```diff
@@ -504,24 +504,122 @@ func (p *ironicProvisioner) ValidateManagementAccess(data provisioner.Management
-		p.debugLog.Info("have active host", "image_source", ironicNode.InstanceInfo["image_source"])
-		return
+		fallthrough
 
 	default:
+		switch data.State {
+		case metal3v1alpha1.StateInspecting,
+			metal3v1alpha1.StatePreparing,
+			metal3v1alpha1.StateProvisioning,
+			metal3v1alpha1.StateDeprovisioning:
+			if deployImageInfo == nil {
+				if p.config.havePreprovImgBuilder {
+					result, err = transientError(provisioner.ErrNeedsPreprovisioningImage)
+				} else {
+					result, err = operationFailed("no preprovisioning image available")
+				}
+			}
+		}
 		return
 	}
 }
 
+// PreprovisioningImageFormats returns a list of acceptable formats for a
+// pre-provisioning image to be built by a PreprovisioningImage object. The
+// list should be nil if no image build is requested.
+func (p *ironicProvisioner) PreprovisioningImageFormats() ([]metal3v1alpha1.ImageFormat, error) {
+	if !p.config.havePreprovImgBuilder {
+		return nil, nil
+	}
+
+	accessDetails, err := p.bmcAccess()
+	if err != nil {
+		return nil, err
+	}
+
+	var formats []metal3v1alpha1.ImageFormat
+	if accessDetails.SupportsISOPreprovisioningImage() {
+		formats = append(formats, metal3v1alpha1.ImageFormatISO)
+	} else {
+		if p.config.deployKernelURL != "" && p.config.deployRamdiskURL != "" {
+			// This is a PXE driver (no ISO support) so it shouldn't require any
+			// network customisation to boot the image, and we have sufficient
+			// data available to configure. Therefore, do not request an image
+			// build.
+			return nil, nil
+		}
+	}
+	if p.config.deployKernelURL != "" {
+		formats = append(formats, metal3v1alpha1.ImageFormatInitRD)
+	}
+
+	return formats, nil
+}
+
+func setDeployImage(driverInfo map[string]interface{}, config ironicConfig, accessDetails bmc.AccessDetails, hostImage *provisioner.PreprovisioningImage) optionsData {
+	deployImageInfo := optionsData{
+		deployKernelKey:  nil,
+		deployRamdiskKey: nil,
+		deployISOKey:     nil,
+	}
+
+	defer func() {
+		// Copy into driverInfo so that if we end up creating a new Node, the
+		// info will be there from the outset.
+		for k, v := range deployImageInfo {
+			if v != nil {
+				driverInfo[k] = v
+			}
+		}
+	}()
+
+	allowISO := accessDetails.SupportsISOPreprovisioningImage()
+	allowInitRD := config.deployKernelURL != ""
+
+	if hostImage != nil {
+		switch hostImage.Format {
+		case metal3v1alpha1.ImageFormatISO:
+			if allowISO {
+				deployImageInfo[deployISOKey] = hostImage.ImageURL
+				return deployImageInfo
+			}
+		case metal3v1alpha1.ImageFormatInitRD:
+			if allowInitRD {
+				deployImageInfo[deployKernelKey] = config.deployKernelURL
+				deployImageInfo[deployRamdiskKey] = hostImage.ImageURL
+				return deployImageInfo
+			}
+		}
+	}
+
+	if !config.havePreprovImgBuilder {
+		if allowISO && config.deployISOURL != "" {
+			deployImageInfo[deployISOKey] = config.deployISOURL
+			return deployImageInfo
+		}
+	}
+	if !config.havePreprovImgBuilder || (hostImage != nil && !allowISO) {
```

Comment:
Re: https://github.com/metal3-io/baremetal-operator/pull/936/files#r714032665 I think this has to change to:

    if !config.havePreprovImgBuilder || !allowISO {

because otherwise it is out of sync with `PreprovisioningImageFormats()`, which does not request a PreprovisioningImage CR to be created if this is a PXE-only host (at line 560).